### PR TITLE
Speed up working of snowplow_page_views model

### PR DESCRIPTION
**Pull request by sphinks**
Ok, I have digged quite a lot of time into the performance issues of the model. Here is my changes. I will comment what was investigated.

I have started with the query plan and checking amount of time model is building on subset of my prod data (about 410-430 seconds). I have to notice that Redshift creates quite a misleading plan, unfortunately. It does not reflect actual complexity of performing some parts of code and `analyse` does not help to fix this issue (at least in my setup). In my initial check of query plan I have notice such part:

```
->  XN Window  (cost=2000024289446.13..2000026736224.05 rows=81559264 width=352)
  Partition: domain_userid
  Order: dvce_created_tstamp
     ->  XN Sort  (cost=2000024289446.13..2000024493344.29 rows=81559264 width=352)
          Sort Key: domain_userid, dvce_created_tstamp
           ->  XN Network  (cost=1000011533028.59..1000013572010.19 rows=81559264 width=352)
             Distribute
             ->  XN Window  (cost=1000011533028.59..1000013572010.19 rows=81559264 width=352)
               Partition: domain_sessionid
               Order: dvce_created_tstamp
                ->  XN Sort  (cost=1000011533028.59..1000011736926.75 rows=81559264 width=352)
                  Sort Key: domain_sessionid, dvce_created_tstamp
                  ->  XN Network  (cost=0.00..815592.64 rows=81559264 width=352)
                   Distribute
                   ->  XN Seq Scan on snowplow_web_events  (cost=0.00..815592.64 rows=81559264 width=352)
```
Just take a look how dramatically cost increased due to sort and window functions. I started from that part of code. It is easy to find that it is reflecting this code: https://github.com/fishtown-analytics/snowplow/blob/master/macros/adapters/default/page_views/snowplow_web_events_internal_fixed.sql

I try to improve sorting keys in `snowplow_web_events` model and it bring a little bit performance in query plan and to actual performance in seconds (I have cuted down model building to 380s). I go further with the `snowplow_web_events_internal_fixed` and turned it to incremental build table instead of `ephemeral` one and get no significant improve of time and query plan. It looks very strange to me.

I have reverted `snowplow_web_events_internal_fixed` and go deeper with analysis of query log in `svl_query_summary` table. And notice very expensive scan of `snowplow_web_events`, unfortunately I can not find corresponding part in the query plan (all scans in plan looks quite cheap comparing to sort and window functions mentioned above). I have analyze all straight `select * from snowplow_web_events`. It leads me to introduction incremental processing of events with some backwindow as I mentioned in original idea of speed up:

```
with all_events as (

    select * from {{ ref('snowplow_web_events') }}
    {% if is_incremental() %}
    where collector_tstamp > (
        DATEADD('day', -1, (select coalesce(max(max_tstamp), '0001-01-01') from {{ this }}))
        )
    {% endif %}
),
...
```

It leads to cutting the total cost in plan quite a big number and I get about 200-220 seconds to build model in my setup. However, I still get a heavy scan of `snowplow_web_events`. I have the only idea how to identify the slow part of SQL - execute it manually part by part. It gives me a great result, I discovered that condition 
```
(a.useragent not {{ snowplow.similar_to('%(bot|crawl|slurp|spider|archiv|spinn|sniff|seo|audit|survey|pingdom|worm|capture|(browser|screen)shots|analyz|index|thumb|check|facebook|PingdomBot|PhantomJS|YandexBot|Twitterbot|a_archiver|facebookexternalhit|Bingbot|BingPreview|Googlebot|Baiduspider|360(Spider|User-agent)|semalt)%') }}
```
requires almost 90% of time within my dataset. Redshift documentation clearly points to the fact `similar to` is very expensive operation and it is better to replace it with `like` expression. I gave it a try and get speed up the code up to 10x times (about 15-18 seconds within my dataset). The strange thing is that it initial plan `similar to` cost quite cheap. Take a look:
```
->  XN Seq Scan on snowplow_web_events  (cost=0.00..2242879.76 rows=1 width=2228)
   Filter: ((((useragent)::text !~ '^(.*(.*(bot|crawl|slurp|spider|archiv|spinn|sniff|seo|audit|survey|pingdom|worm|capture|(browser|screen)shots|analyz|index|thumb|check|facebook|PingdomBot|PhantomJS|YandexBot|Twitterbot|a.archiver|facebookexternalhit|Bingbot|BingPreview|Googlebot|Baiduspider|360(Spider|User-agent)|semalt).*).*)$'::text) OR (useragent IS NULL)) AND (collector_tstamp > date_add('hour'::text, -1::bigint, $0)) AND ((COALESCE(br_type, 'unknown'::character varying))::text <> 'Bot/Crawler'::text) AND ((COALESCE(br_type, 'unknown'::character varying))::text <> 'Robot'::text) AND (((br_family)::text <> 'Robot/Spider'::text) OR (br_family IS NULL)) AND (domain_sessionidx > 0) AND (domain_userid IS NOT NULL))
```
Compare it to the cost of code `snowplow_web_events_internal_fixed`. Real result is different.
In total I get almost 10x performance increase within my prod data: from 1800 seconds to 200 seconds.

### CODE REVIEW

**Comment by sphinks** [comment]
Ok, I have digged quite a lot of time into the performance issues of the model. Here is my changes. I will comment what was investigated.

I have started with the query plan and checking amount of time model is building on subset of my prod data (about 410-430 seconds). I have to notice that Redshift creates quite a misleading plan, unfortunately. It does not reflect actual complexity of performing some part of code and `analyse` does not help to fix this issue (at least in my setup). In my initial check of query plan I have notice such part:

```
->  XN Window  (cost=2000024289446.13..2000026736224.05 rows=81559264 width=352)
  Partition: domain_userid
  Order: dvce_created_tstamp
     ->  XN Sort  (cost=2000024289446.13..2000024493344.29 rows=81559264 width=352)
          Sort Key: domain_userid, dvce_created_tstamp
           ->  XN Network  (cost=1000011533028.59..1000013572010.19 rows=81559264 width=352)
             Distribute
             ->  XN Window  (cost=1000011533028.59..1000013572010.19 rows=81559264 width=352)
               Partition: domain_sessionid
               Order: dvce_created_tstamp
                ->  XN Sort  (cost=1000011533028.59..1000011736926.75 rows=81559264 width=352)
                  Sort Key: domain_sessionid, dvce_created_tstamp
                  ->  XN Network  (cost=0.00..815592.64 rows=81559264 width=352)
                   Distribute
                   ->  XN Seq Scan on snowplow_web_events  (cost=0.00..815592.64 rows=81559264 width=352)
```
Just take a look how dramatically cost increased due to sort and window functions. I started from that part of code. It is easy to find that it is reflecting this code: https://github.com/fishtown-analytics/snowplow/blob/master/macros/adapters/default/page_views/snowplow_web_events_internal_fixed.sql

I try to improve sorting keys in `snowplow_web_events` model and it bring a little bit performance in query plan and to actual performance in seconds (I have cut down model building to 380s). I go further with the `snowplow_web_events_internal_fixed` and turned it to incremental build table instead of `ephemeral` one and get no significant improve of time and query plan. It looks very strange to me.

I have reverted `snowplow_web_events_internal_fixed` and go deeper with analysis of query log in `svl_query_summary` table. And notice very expensive scan of `snowplow_web_events`, unfortunately I can not find corresponding part in the query plan (all scans in plan looks quite cheap comparing to sort and window functions mentioned above). I have analyze all straight `select * from snowplow_web_events`. It leads me to introduction incremental processing of events with some backwindow as I mentioned in original idea of speed up:

```
with all_events as (

    select * from {{ ref('snowplow_web_events') }}
    {% if is_incremental() %}
    where collector_tstamp > (
        DATEADD('day', -1, (select coalesce(max(max_tstamp), '0001-01-01') from {{ this }}))
        )
    {% endif %}
),
...
```

It leads to cutting the total cost in plan quite a big number and I get about 200-220 seconds to build model in my setup. However, I still get a heavy scan of `snowplow_web_events`. I have the only idea how to identify the slow part of SQL - execute it manually part by part. It gives me a great result, I discovered that condition 
```
(a.useragent not {{ snowplow.similar_to('%(bot|crawl|slurp|spider|archiv|spinn|sniff|seo|audit|survey|pingdom|worm|capture|(browser|screen)shots|analyz|index|thumb|check|facebook|PingdomBot|PhantomJS|YandexBot|Twitterbot|a_archiver|facebookexternalhit|Bingbot|BingPreview|Googlebot|Baiduspider|360(Spider|User-agent)|semalt)%') }}
```
requires almost 90% of time within my dataset. Redshift documentation clearly points to the fact `similar to` is very expensive operation and it is better to replace it with `like` expression. I gave it a try and get speed up the code up to 10x times (about 15-18 seconds within my dataset). The strange thing is that it initial plan `similar to` cost quite cheap. Take a look:
```

**Comment by sphinks** [comment]
@drewbanin really want to hear your opinion regarding my changes and new snowplow script releases, to not going into custom branch.

**Comment by drewbanin** [comment]
@sphinks this is fascinating! I had no idea that `similar_to` was so slow on Redshift!! I do think the logic you're using in the `useragent` filter is backwards though -- we want to _exclude_ the events which come from know bot useragents. Is it possible that downstream models are running faster because the code you're using filters a much larger set of input events? Regardless, I buy that we should use `like` instead of `similar to` here.

I think we probably shouldn't use this full set of sort keys in `snowplow_web_events`: `['page_view_id', 'collector_tstamp', 'domain_sessionid', 'domain_userid', 'dvce_created_tstamp']`

In `snowplow_page_views`, you've added:

```
with all_events as (
    select * from {{ ref('snowplow_web_events') }}
    {% if is_incremental() %}
    where collector_tstamp > (
        DATEADD('day', -1, (select coalesce(max(max_tstamp), '0001-01-01') from {{ this }}))
        )
    {% endif %}
```

I like this approach a lot! The key thing here is that we're filtering on the `collector_tstamp` field of `snowplow_web_events`, so we'll want to make sure that `collector_tstamp` is the first sort key in the `snowplow_web_events` table. This will help Redshift limit the amount of data that it needs to scan to build `snowplow_page_views`. I'd be super curious hear how performance fares if we make the sort key `collector_tstamp` instead of `page_view_id`.

Nice work @sphinks!! Thanks so much for opening up this PR, I'm very keen to discuss it :)

**Comment by sphinks** [comment]
> I do think the logic you're using in the useragent filter is backwards though -- we want to exclude the events which come from know bot useragents.

It is totally equal to the previous version of code in this part.

> Is it possible that downstream models are running faster because the code you're using filters a much larger set of input events?

I do not compare the performance of downstream models, so not sure. Useragent filter does not filter more event, it filters the same amount of events, just executing much faster. The filter of the incoming event into the model page_views is affecting only page_view model and does not connected with the downstream model. Could you clarify what filter are you expecting will increase the performance of downstream models?

> I think we probably shouldn't use this full set of sort keys in snowplow_web_events: ['page_view_id', 'collector_tstamp', 'domain_sessionid', 'domain_userid', 'dvce_created_tstamp']

I do not have strong facts if that way is better, just discover some recommendation use up to 5 sort fields is recommended, so I put all fields that are used in filtering or joins.

> I like this approach a lot! The key thing here is that we're filtering on the collector_tstamp field of snowplow_web_events, so we'll want to make sure that collector_tstamp is the first sort key in the snowplow_web_events table. This will help Redshift limit the amount of data that it needs to scan to build snowplow_page_views. I'd be super curious to hear how performance fares if we make the sort key collector_tstamp instead of page_view_id.

Correct! We can put collector_tstamp as the first sort key, however, I do not see much difference in my experiments with sort keys. It becomes faster several seconds that could be not the performance gain but just some other facts. So we can easily switch it but I can not say it will increase performance. Looks like Redshift is using composite sort key quite effective to avoid huge performance lags due to field order. I left the page_view_id in the first place as it was because do not want to affect the performance of other models.

**Comment by drewbanin** [comment]
I don't think that's right -- the previous logic reads:
```
and (a.useragent not {{ snowplow.similar_to('...') }} ....
```
whereas the logic you've implemented reads:
```
and (useragent like '%bot%' or useragent like '%crawl%' or useragent like '%slurp%' ...)
```

So, previously this logic would accept events which did not have a bot useragent. This new logic only accepts events which have a bot useragent. Let me know if I'm missing something here.

Duly noted on the sort key front -- that all makes a lot of sense

**Comment by sphinks** [comment]
> whereas the logic you've implemented reads:
 
OMG. Good catch! I forget to add `not` before like condition. Pushed changes for correct like condition and implement new order of sort keys, just put collector_tstamp on a first place (as I do not see significant difference running the whole model, so let it be on a first place).

File: macros/adapters/default/page_views/snowplow_page_views.sql [hunk 1]
```diff
@@ -24,9 +24,16 @@
...
 {% set use_useragents = (var('snowplow:context:useragent') != false) %}
 
+-- we are using 1-days window allowing us cover most of sessions
+-- but model become much faster comparing to selecting all events
 with all_events as (
 
     select * from {{ ref('snowplow_web_events') }}
+    {% if is_incremental() %}
+    where collector_tstamp > (
+        DATEADD('day', -1, (select coalesce(max(max_tstamp), '0001-01-01') from {{ this }}))
```
Comment (jtcohen6) [inline macros/adapters/default/page_views/snowplow_page_views.sql]: A 1-day "look-back" is reasonable here, since the vast majority (~99%) of events will arrive within 24 hours of firing. I think we should make the interval length a configurable variable, so that analysts can choose 3, 7, or 30 days if they find that some quirk of their users or collector setup requires a greater "look-back" window.

That would look like:
```
where collector_tstamp > (
        DATEADD('day', -1 * {{var('page_view_lookback_days')}}, (select coalesce(max(max_tstamp), '0001-01-01') from {{ this }}))
```
Where `page_view_lookback_days` is defined as a variable in `dbt_project.yml` with a default value. My personal instinct for the default is to look back 3 days, but open to your thoughts!

File: macros/adapters/default/page_views/snowplow_web_events.sql [hunk 2]
```diff
@@ -15,7 +15,7 @@
 {{
     config(
         materialized='incremental',
-        sort='page_view_id',
+        sort=['collector_tstamp', 'page_view_id', 'domain_sessionid', 'domain_userid', 'dvce_created_tstamp'],
```
Comment (jtcohen6) [inline macros/adapters/default/page_views/snowplow_web_events.sql]: I think a compound sort key on `collector_tstamp` + `page_view_id` makes sense. Could you make this change in a separate PR? I understand it's motivated by `snowplow_page_views` performance, but I think it's outside the scope of tweaking the model SQL of `snowplow_page_views` in particular.

**Review by jtcohen6 — COMMENTED** [review]
@sphinks I just had a chance to catch up with @drewbanin on our ambitions for this package. The core change in this PR is looking good. I left a few comments, mainly around adding a user-configurable variable for "look-back" window in `snowplow_page_views`.

**Comment by sphinks** [comment]
@jtcohen6 I have introduce var for lookback days with default value and reverted sort keys list. Will create PR on sort keys, once this one will be merged as I have to revert my forked branch to avoid appearing change from current PR there.

**Comment by jtcohen6** [comment]
@sphinks Nice work!

I see you added the default value in a slightly different manner from what I had imagined. I was thinking we would include the var in `dbt_project.yml` like so:
```yml
models:
  snowplow:
  ...
    vars:
      ...
      'snowplow:pass_through_columns': []
      'snowplow:page_view_lookback_days': 1
```
As well as updating the package README [here](https://github.com/fishtown-analytics/snowplow#configuration).

I think this approach is preferable since it makes the configuration more visible to the end user.

**Comment by sphinks** [comment]
@jtcohen6 done!
However, my initial idea was that such setting is not so important for a new users and they want as few as variables to set as possible, so it has default value. Experienced users could find new setting from documentation and code. Not sure if I'm correct. Nevertheless, I suppose you have better understanding of DBT users.

**Comment by jtcohen6** [comment]
@sphinks That totally makes sense! For now, I think I'd rather err on the side of exposing more available configuration rather than less. There are several ways in which we want to improve the user experience for this package, and we might revisit then.

I kicked off tests. It looks like they failed because you need to add the default value
```yml
'snowplow:page_view_lookback_days': 1
```
to the `dbt_project.yml` file itself.

**Comment by sphinks** [comment]
@jtcohen6 Opps, I missed the face that it is existing in snowplow modules, not only in user projects. Add it. Please, try once again with tests.

**Review by jtcohen6 — APPROVED** [review]
@sphinks Looks like all our tests passed! I'll merge your PR now, for inclusion in our next minor release (`0.8.0`).

**Comment by sphinks** [comment]
@jtcohen6 sorry I was on vacation. Now I'm back. When we can expect 0.8.0?

**Comment by jtcohen6** [comment]
@sphinks Thanks for the reminder on this. There are a few other open issues + PRs, but I don't think those should block all the work we've already merged into `0.8.0`. I'll cut a release now.
